worldedit-core/src/main/java/com/sk89q/worldedit/command/tool/brush/ImageBrush.java

```diff
@@ -0,0 +1,110 @@
+    private final boolean fill;
+    private final boolean flat;
+
+    public ImageBrush(BufferedImage image, double intensity, boolean fill, boolean flat) {
+        this.image = image;
+        this.intensity = intensity;
+        this.fill = fill;
+        this.flat = flat;
+    }
+
+    @Override
+    public void build(EditSession editSession, BlockVector3 position, Pattern pattern, double doubleSize) throws MaxChangedBlocksException {
+        // Resize the image
+        int size = (int) Math.ceil(doubleSize);
+        int diameter = size * 2 + 1;
+        BufferedImage resizedPattern = new BufferedImage(diameter, diameter, 1);
+        Graphics2D graphic = null;
+        try {
+            graphic = resizedPattern.createGraphics();
+            graphic.drawImage(this.image, 0, 0, diameter, diameter, null);
+        } finally {
+            graphic.dispose();
+        }
+
+        double random = RANDOM.nextDouble();
+        for (int offX = -size; offX <= size; offX++) {
+            for (int offZ = -size; offZ <= size; offZ++) {
+                int posX = position.getX() + offX;
+                int posZ = position.getZ() + offZ;
+                int posY = editSession.getHighestTerrainBlock(posX, posZ, 0, 255, editSession.getMask());
+                BlockVector3 block = BlockVector3.at(posX, posY, posZ);
+                if (editSession.getMask() != null && !editSession.getMask().test(block)) {
+                    continue;
+                }
+
+                double height = getHeightAt(resizedPattern, offX + size, resizedPattern.getHeight() - 1 - (offZ + size));
+                // Add a bit of variation
+                if (random > 1 - height % 1) {
+                    height += 1;
+                }
+
+                BaseBlock baseBlock = fill ? editSession.getBlock(block).toBaseBlock() : null;
+                for (int y = 0; y < height; y++) {
+                    if (fill) {
+                        // Only go up to the origin's level if flat mode is enabled
+                        if (!flat || block.getY() + y <= position.getY()) {
+                            editSession.setBlock(block.withY(block.getY() + y), baseBlock);
+                        }
+                    } else if ((!flat || block.getY() - y > position.getY()) && block.getY() - y >= 0
+                            && (editSession.getMask() == null || editSession.getMask().test(block.withY(block.getY() - y)))) {
+                        // Remove blocks if not using fill mode
+                        editSession.setBlock(block.withY(block.getY() - y), BlockTypes.AIR.getDefaultState());
+                    }
+                }
+            }
+        }
+    }
+
+    private double getHeightAt(BufferedImage image, int x, int y) {
+        int rgb = image.getRGB(x, y);
+
+        int red = rgb >>> 16 & 0xFF;
+        int green = rgb >>> 8 & 0xFF;
+        int blue = rgb & 0xFF;
+        if (red == 0 && green == 0 && blue == 0) {
```

Comment:
Is there a reason for this short circuit?

I know I'm probably nit-picking here, but I feel that this would be "slower" than not having it, due to extra complexity preventing JVM optimisations.